### DISCUSSION TITLE
Will SimpleReconGadget do Channels Combination? And what does E0, E1 , E2, CHA, N, S and LOC mean?

### DISCUSSION THREAD

**Discussion by zhengliuX** [gadgetron/gadgetron#1192] [Q&A] [answered]
Hi, I'm using gadgetron to do mri reconstruction.
I want to know if ```SimpleReconGadget``` will do channels combiantion, and I noticed this code.
May I ask what do ```E0```, ```E1```, ```E2```, ```CHA```, ```N```, ```S``` and ```LOC``` really mean? And is it OK if I directly transfer a complex image without channels combination to ```SimpleReconGadget```?

https://github.com/gadgetron/gadgetron/blob/86d954bf9f012c407b0cc94b98af659b14e59ab3/gadgets/mri_core/SimpleReconGadget.cpp#L20-L27

**Comment by zhengliuX** [comment]
@hansenms  Hi, any ideas?

**Comment by hansenms** [comment]
The SimplreReconGadget uses an internal Gadgetron data structure defined here https://github.com/gadgetron/gadgetron/blob/master/toolboxes/mri_core/mri_core_data.h using this data structure, we say that all recon problems are at most 7 dimensional. E0, E1, E2 refers to kx, ky, kz dimension sizes, CHA, is number of channels, LOC is location (slice or volume), and N and S are to variable dimensions, they could be time and contrast or some other choices. We use some Gadgets like the accumulator (which produces buckets of data) and BucketToBuffer, which produces this recondata data structure from these buckets.

I am not sure what you mean by whether it is OK to transfer complex images without channel combination. Uncombined complex images *could* be held in a recondata data structure, but since they would already be in image space, I am not sure it makes much sense to send them to the recon gadget.

Not sure if this helps, but I will need more information about what you are trying to do if not.

**Reply by Andrew-Dupuis** [reply]
@hansenms I definitely feel like the issue of "what is what dimension" in the recondata space is a barrier for new users, especially as the data flows further and further down the chain. Is there an accessor method anywhere that tracks/ allows you to get the mapping for what's been pushed into the N and S dimensions? This has definitely tripped me up, as well as many folks in our lab. 

As an interim, we should also take the explanation you wrote above and add it (somewhere) to the documentation.

**Comment by Andrew-Dupuis** [comment]
@ZhengLiu97 - to add to Michael's answer quickly, SimpleReconGadget makes a lot of assumptions about the kind of data and desired processing. Unless your use case matches pretty specifically, its not likely to be the right Gadget for you.  Also, SimpleReconGadget performs the FFT to image space - I'm not sure why you'd be passing a complex image into the gadget. The data pre-input to that gadget  should be complex kspace information. Echoing Michael, having some more details about your use case would help a ton in specc'ing out an initial pipeline for you! Many of us are at ISMRM in Sedona right now, so we may be a bit slow to respond until later in the week - but you're welcome to join the Gadgetron user meeting on Friday at 10am EST (links are pinned in the discussion) and we can chat a bit more!

**Comment by zhengliuX** [comment]
Thanks you so much for the reply and explanation. @hansenms @Andrew-Dupuis 

As to the dimension of the Gadgetron data structure, I still have little confusion. As I know, the ```kx``` is the frequency(or readout) dimension of the kspace, ```ky``` is the encoding dimension, and the ```kz``` is the slice dimension. If so, does the ```LOC``` repeats the semantic of ```kz```? Or I just misunderstood something here. 

And I did make some mistakes when describing the question about passing the complex image to a gadget. Yes, I actually not passed the to the SimpleReconGadget , Here is the situation,  what I want to do is reconstructing the channel-combined complex image using python ( as well as its xml config file), and pass it to a writer gadget, saving a .h5 file. The  problem I'm facing is, I know exactly how to do channel combination when passing a amplitude image, to be short, RMS of the absolute value of the complex images, however, how to do channel combination on the complex images, and return **a one-channel complex image** is really confusing.

I've read code from ```CombineGadget```, it combines multi-channel complex images using these algorithm,
https://github.com/gadgetron/gadgetron/blob/61fb5481815fbb7b369c4b8fdd7de2ba7907e0d2/gadgets/mri_core/CombineGadget.cpp#L43-L57
Could you please help me understand why calculating phase this way, and combine mag and phase into a new complex data?

**Comment by hansenms** [comment, marked as answer]
On your first question, think of LOC as excited volume. So if you have a 2D multi-slice (not 3D) acquisition, it would be slice, if you have a 3D acquisition, typically you only have one location, but you could have an acquisition acquiring multiple 3D volumes and then LOC would index each volume.

As for coil combination to produce a complex image with useable phase, that is a complex topic (pun intended). In the algorithm above, the phase in the combined image would be a weighted sum of the phases in each channel. The thinking is that is the magnitude is low (the phase is random), we should weight that phase contribution low, etc.

The "right" way to do it is by using proper coil sensitivity maps that have phase too and use standard coil combination. However, getting to those coil sensitivity maps is not straight forward, there are approximate algorithms such as the Walsh or the Inati algorithm (both have implementations in Gadgetron), so you could start there.

I would recommend you send me your email and come to the Friday meetings and discuss your problem, it will be a lengthy correspondence here.

**Comment by zhengliuX** [comment]
Thank you so much!
Really, great help！

And I just sent you my email, hope you received it. Also, what software should I download? Zoom or Team?

**Comment by hansenms** [comment]
@ZhengLiu97 I don't see any email from you. I have checked multiple mailboxes including spam. Please email at firstname dot lastname AT microsoft.com.

**Reply by zhengliuX** [reply]
@hansenms Sorry, I noticed that I did send to a wrong email address, I just sent another email called ```Email for Gadgetron Weekly Hangout```. I cannot check if the email was successfully sent for I'm using Outlook, so please inform me if you don't get my email.
